Delegated Credentials (WORK IN PROGRESS)

tsusanka opened the pull request:
This is not yet finished, this PR is more of a "checkpoint".

TODO: 
- [x] saving the credentials to use them until they're valid
- [ ] test edge cases
- [ ] TLS 1.3 support
- [ ] incorporate into keyless

cc @grittygrease

grittygrease commented on handshake_server.go:
```diff
@@ -356,6 +369,24 @@ Curves:
 		}
 	}
 
+	if hs.clientHello.delegatedCredentials {
```
This logic should most likely be moved fully into GetCertificate. A cleaner design would be to simply create a new PrivateKey in the certificate object that corresponds to the delegated credential private key. That way almost all logic on the server side would be unchanged, aside from the addition of a new opaque delegated credentials object in the serverhello.

tsusanka replied:
Thanks. I'm a little confused, how to use the GetCertificate function. If the c.GetCertificate is not null it is called to retrieve the certificate.

So, my goal is to implement a new GetCertificate function which I then "override" how the certificate is retrieved, right? The function will add the delegated credential and its PrivateKey. However, where this code should be? Also in the subpackage? Where do I then set the config to use this function?

tsusanka replied:
Ah, alright, I simply set the function in the config when spinning up TLS, right?

grittygrease commented on handshake_delegated_cred.go:
```diff
@@ -0,0 +1,177 @@
+	"errors"
+	"time"
+)
+
```
This tight integration with the handshake is not ideal. I think all the credential minting and verifying should be abstracted into its own package (byte slice + cert to PrivateKey and back). The only pieces that need to change on the client side are the sending of the empty extension, the extraction and validation of the delegated credential, and the CertificateVerify function.

tsusanka replied:
>  The only pieces that need to change on the client side are the sending of the empty extension, the extraction and validation of the delegated credential, and the CertificateVerify function.

I'm confused about the CertificateVerify part. Do you mean the CertificateVerify message? But that's used for client authentication, isn't it?

grittygrease replied:
Actually, in TLS 1.2 this is called the ServerKeyExchange.

grittygrease commented on common.go:
```diff
@@ -11,6 +11,7 @@ import (
 	"crypto/rand"
 	"crypto/sha512"
 	"crypto/x509"
+	"encoding/asn1"
```
As mentioned below, this is not a great dependency to add. Factor out as much logic as possible into its own package and try to keep the changes to crypto/tls minimal.

grittygrease commented on common.go:
```diff
@@ -949,6 +954,9 @@ type Certificate struct {
 	OCSPStaple []byte
+	// DelegatedCredential contains a signed object containing a public
+	// key for signing handshakes
+	DelegatedCredential DelegatedCredential
```
This is probably the wrong place for this type of structure. When the Certificate is returned from GetCertificate it should already have a serialized opaque delegated credential, not a structure that needs to be serialized.

tsusanka replied:
I understand. The Certificate struct will then contain marshalled DC and the associated PrivateKey. Both received from the delcred sub package.

grittygrease replied:
Perhaps consider reusing the existing PrivateKey struct.

grittygrease commented on handshake_delegated_cred_test.go:
```diff
@@ -0,0 +1,283 @@
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
```
Again, this is too tightly integrated with the handshake. Each individual function should be unit tested with respect to its functionality, the internal tls structures should not be involved.

tsusanka replied:
Okay. So I should add unit tests - e.g. check if the credential marshalled correctly etc. What is wrong with those higher-level tests though? They still provide value, don't they?

grittygrease commented:
Config.Clone and other ancillary functions are also missing the new data.

grittygrease commented:
Also note that the tests are failing.

tsusanka commented on common.go:
```diff
@@ -614,6 +616,9 @@ type Config struct {
 	originalConfig *Config
+
+	// states if Delegated Credentials should be used
+	useDelegatedCredentials bool
```
You're right, however, how come I still can access it? https://github.com/tsusanka/tls-tris/blob/tsusanka/delegated-cred/handshake_delegated_cred_test.go#L254

tsusanka commented on common.go:
```diff
@@ -959,6 +967,23 @@ type Certificate struct {
+
+// checks certificate if it contains the DelegationUsage extension
+// required for Delegated Credentials
+func (hs *serverHandshakeState) isCertificateValidForDelegationUsage() bool {
```
Got it. The content of the handshake_delegated_cred.go file will be then moved to separate crypto/tls/delcred (or similar name) sub package. This sub package will be responsible for creating, marshalling, signing, validating etc. of the credentials.

grittygrease replied:
Yes, that sounds like a plan.

tsusanka commented:
Thanks for the review. I'm aware of the failing tests, it is some kind of race condition issue, I'll have a look into that once I fix the outlined issues.

tsusanka commented on cipher_suites.go:
```diff
@@ -33,7 +34,7 @@ type keyAgreement interface {
 	// This method may not be called if the server doesn't send a
 	// ServerKeyExchange message.
-	processServerKeyExchange(*Config, *clientHelloMsg, *serverHelloMsg, *x509.Certificate, *serverKeyExchangeMsg) error
+	processServerKeyExchange(*Config, *clientHelloMsg, *serverHelloMsg, crypto.PublicKey, *serverKeyExchangeMsg) error
```
It might be possible to avoid this change by wrapping the DC's PublicKey into an x509.Certificate. However:
a) I'm not sure it'll work since we have only the PublicKey
b) all processServerKeyExchange functions use only the public key, so the dependency for x509.Certificate seems  redundant
c) the function is not exported, so it should not be a BC break (?)

grittygrease replied:
This is a slick change. I can't see why it wouldn't work or why *x509.Certificate was used in the first place.

tsusanka commented:
I think this is ready for another round of code review.
- All code is in the same _tls_ package, however, the logic is clearly separated using the GetCertificate function as discussed
- I've squashed all commits into just a few, it was a mess
- Tests are fixed
- I've implemented the support for PKCS1, it's trivial to remove it in case we decide to

grittygrease commented on delegated_credentials.go:
```diff
@@ -0,0 +1,250 @@
+}
+
+func NewDelCredGetCertificateFunction(cert *Certificate) GetCertificate {
+
```
Unnecessary space?

grittygrease commented on delegated_credentials.go:
```diff
@@ -0,0 +1,250 @@
+	PublicKey interface{}
+}
+
+func NewDelCredGetCertificateFunction(cert *Certificate) GetCertificate {
```
This abbreviation is a bit awkward. NewGetCertificateFromDelegatedCredential? That's also long but a bit better.

tsusanka replied:
Yeah, I don't really like it myself. I'm not sure the `From` is accurate. So how about simply `NewDelegatedCredentialGetCertificate`?

grittygrease commented on delegated_credentials.go:
```diff
@@ -0,0 +1,250 @@
+func NewDelCredGetCertificateFunction(cert *Certificate) GetCertificate {
+
+	return func(clientHelloInfo *ClientHelloInfo) (*Certificate, error) {
+
```
Space

grittygrease commented on delegated_credentials.go:
```diff
@@ -0,0 +1,250 @@
+}
+
+func selectVersion(versions []uint16) uint16 {
+
```
Newline

grittygrease commented on delegated_credentials.go:
```diff
@@ -0,0 +1,250 @@
+	if scheme == ECDSAWithP256AndSHA256 {
+		privateKey, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		credential.PublicKey = privateKey.(crypto.Signer).Public()
+
```
newline

grittygrease commented on delegated_credentials.go:
```diff
@@ -0,0 +1,250 @@
+
+	copy(cred[7:7+publicKeyLength], publicKeyBytes)
+
+	cred[publicKeyLength+7] = uint8(scheme >> 8) // hash
```
This will likely be different for TLS 1.3

grittygrease commented on handshake_client.go:
```diff
@@ -21,13 +21,14 @@ import (
+	finishedHash     finishedHash
+	masterSecret     []byte
+	session          *ClientSessionState
+	delCredPublicKey crypto.PublicKey
```
Maybe this should be just connection PublicKey, and it is set to either the delegated cred key or c.peerCertificates[0].PublicKey. Then you may be able to reduce the number of if statements below.

tsusanka commented on delegated_credentials.go:
```diff
@@ -0,0 +1,295 @@
+)
+
+// After only renewThreshold of credential's validTime is left a new one is created
+const renewThreshold = 0.2
```
Maybe better to reverse this? Set it to `0.8` and call it `renewAfterThreshold`

grittygrease commented on delegated_credentials.go:
```diff
@@ -0,0 +1,295 @@
+// SignatureScheme to ensure client's support.
+func createDelegatedCredential(config *CredentialsConfig) (credential DelegatedCredential, privateKey crypto.PrivateKey, renewAfter time.Time, err error) {
+	validTill := config.timeFunc().Add(config.Validity)
+	relativeToCert := validTill.Sub(config.Cert.Leaf.NotBefore)
```
The computation of the ValidTime is a rather complex dance. Is there a simpler way to compute it that is more readable?

tsusanka replied:
Yeah, I agree. I was thinking about it and I'm not sure there is a better way though. The code is basically the definition in itself:
> Relative time in seconds from the beginning of the certificate's notBefore value after which the Delegated Credential is no longer valid

Btw why is this relative? Why not making it a simple datetime when the credential expires? The client needs to verify the certificate's validity anyway.

----

I can remove the `relativeToCert` variable but it's not significantly better:

```go
validTill := config.timeFunc().Add(config.Validity)
renewAfter = validTill.Add(-time.Duration(config.Validity.Seconds()*renewThreshold) * time.Second)
credential = DelegatedCredential{
	ValidTime: int64(validTill.Sub(config.Cert.Leaf.NotBefore).Seconds()),
}
```

grittygrease commented on delegated_credentials.go:
```diff
@@ -0,0 +1,295 @@
+	renewChan chan struct{}
+}
+
+var cachedCredentials map[SignatureScheme]*cachedCredential
```
This is a package-global variable indexed by the SignatureScheme. This means it will be shared between all instantiations, even those with different certs. You most likely want to keep this local to the CredentialsConfig.

tsusanka replied:
Good point! So I'm just including it as a not exported CredentialsConfig variable.

grittygrease commented on delegated_credentials.go:
```diff
@@ -0,0 +1,295 @@
+			go func() {
+				cachedCredentials[config.scheme], err = newCredential(config)
+				if config.renewChan != nil {
+					config.renewChan <- struct{}{}
```
Where is the renewChan read?

tsusanka replied:
Only in [tests](https://github.com/cloudflare/tls-tris/pull/29/files#diff-74969c972e0630d3c7b131f8f82325e9R270). I was reluctant to add the channel just because of tests, but I wanted to test that the new credential is created, which I can do only after the goroutine finishes. I was talking to Salmen about this and he advised me this.

tsusanka commented:
Closed in favour #32
